### ISSUE TITLE
8 RESOURCE_EXHAUSTED: Received message larger than max

### ISSUE THREAD

**Issue by davidfarinha** [konsumer/grpc-dynamic-gateway#38]
Hello,

I'm getting this error `8 RESOURCE_EXHAUSTED: Received message larger than max (7479039 vs. 4194304)` when sending a large request (base64 encoded input string) to my grpc gateway using this library.
I can see this is due to a exceeding a default request size value (of 4MB) set at the grpc server level and I need to pass options to the grpc server as mentioned in [this thread](https://github.com/tensorflow/serving/issues/1382#issuecomment-503384623) (eg `options = [('grpc.max_receive_message_length', max_message_length)]`) to override and increase this request size option to fix this issue. However how can I pass options to the grpc server using this library?
Thanks!

**Comment by konsumer** [comment]
There is a `grpc` param to the middleware that let's you optionally pass your own grpc, but defaults to the import of the `grpc` module.  Without testing, I imagine you could do this:

```js
const grpcGateway = require('grpc-dynamic-gateway')
const express = require('express')
const bodyParser = require('body-parser')
const grpc = require('grpc')

const app = express()
app.use(bodyParser.json())
app.use(bodyParser.urlencoded({ extended: false }))

// DO STUFF TO grpc HERE

// load the proxy on / URL
app.use('/', grpcGateway(
  ['api.proto'],
  '0.0.0.0:5051',
  grpc.credentials.createInsecure(),
  true,
  process.cwd(),
  grpc
))

const port = process.env.PORT || 8080
app.listen(port, () => {
  console.log(`Listening on http://0.0.0.0:${port}`)
})
```
I'm not quite sure how they are using `grpc.insecure_channel`, though, so you might have to play around with it.

**Comment by konsumer** [comment]
The place I am getting my client is [here](https://github.com/konsumer/grpc-dynamic-gateway/blob/master/index.js#L52), where `protos` is a bunch of `@grpc/proto-loader` clients, so I might need to add some options or something. It doesn't look like `@grpc/proto-loader` has any way to set `grpc.max_receive_message_length`, but maybe you can set it globally:

```js
grpc.max_receive_message_length = max_message_length
```
I looked at the grpc object and didn't see this in it, though, so I'm not sure:

```js
{
  loadObject: [Function: loadObject],
  load: [Function: deprecated],
  loadPackageDefinition: [Function: loadPackageDefintion],
  setLogger: [Function: setLogger],
  setLogVerbosity: [Function: setLogVerbosity],
  Server: [Function: Server],
  Metadata: [Function: Metadata] { _fromCoreRepresentation: [Function] },
  status: {
    OK: 0,
    CANCELLED: 1,
    UNKNOWN: 2,
    INVALID_ARGUMENT: 3,
    DEADLINE_EXCEEDED: 4,
    NOT_FOUND: 5,
    ALREADY_EXISTS: 6,
    PERMISSION_DENIED: 7,
    RESOURCE_EXHAUSTED: 8,
    FAILED_PRECONDITION: 9,
    ABORTED: 10,
    OUT_OF_RANGE: 11,
    UNIMPLEMENTED: 12,
    INTERNAL: 13,
    UNAVAILABLE: 14,
    DATA_LOSS: 15,
    UNAUTHENTICATED: 16
  },
  propagate: {
    DEADLINE: 1,
    CENSUS_STATS_CONTEXT: 2,
    CENSUS_TRACING_CONTEXT: 4,
    CANCELLATION: 8,
    DEFAULTS: 65535
  },
  callError: {
    OK: 0,
    ERROR: 1,
    NOT_ON_SERVER: 2,
    NOT_ON_CLIENT: 3,
    ALREADY_INVOKED: 5,
    NOT_INVOKED: 6,
    ALREADY_FINISHED: 7,
    TOO_MANY_OPERATIONS: 8,
    INVALID_FLAGS: 9,
    INVALID_METADATA: 10,
    INVALID_MESSAGE: 11,
    NOT_SERVER_COMPLETION_QUEUE: 12,
    BATCH_TOO_BIG: 13,
    PAYLOAD_TYPE_MISMATCH: 14
  },
  writeFlags: { BUFFER_HINT: 1, NO_COMPRESS: 2 },
  logVerbosity: { DEBUG: 0, INFO: 1, ERROR: 2 },
  methodTypes: {
    UNARY: 0,
    CLIENT_STREAMING: 1,
    SERVER_STREAMING: 2,
    BIDI_STREAMING: 3
  },
  connectivityState: {
    IDLE: 0,
    CONNECTING: 1,
    READY: 2,
    TRANSIENT_FAILURE: 3,
    SHUTDOWN: 4
  },
  credentials: {
    createSsl: [Function],
    createFromMetadataGenerator: [Function],
    createFromGoogleCredential: [Function],
    combineChannelCredentials: [Function],
    combineCallCredentials: [Function],
    createInsecure: [Function]
  },
  ServerCredentials: [Function: ServerCredentials] {
    createSsl: [Function],
    createInsecure: [Function]
  },
  makeGenericClientConstructor: [Function],
  getClientChannel: [Function],
  waitForClientReady: [Function],
  StatusBuilder: [Function: StatusBuilder],
  ListenerBuilder: [Function: ListenerBuilder],
  RequesterBuilder: [Function: RequesterBuilder],
  InterceptingCall: [Function: InterceptingCall],
  closeClient: [Function: closeClient],
  Client: [Function: Client],
  Channel: [Function: Channel]
}
```

**Comment by davidfarinha** [comment]
Appreciate the informative response, very nice to know there is a way to pass options to the grpc server-and I was able to pass options in the way you described. Thanks again!

**Comment by adambeer** [comment]
@davidfarinha Would you be able to tell me how you got this solved? I tried adding `grpc.max_receive_message_length = max_message_length` just before the app.use for grpc gateway and that didnt solve it. How did you set this option?

**Comment by adambeer** [comment]
@konsumer Would you happen to know how this may have been resolved?

**Comment by konsumer** [comment]
It looks like above they set a global setting like this:

```
grpc.max_receive_message_length = max_message_length
```

But I am not sure. It may have changed in the upstream grpc lib.

**Comment by adambeer** [comment]
I figured it out. For anyone who comes across this in the future, I modified the package directly. In index.js on line 77, change `getPkg(clients, pkg, true)[svc] = new (getPkg(protos[si], pkg, false))[svc](grpcLocation, credentials)`

to

`getPkg(clients, pkg, true)[svc] = new (getPkg(protos[si], pkg, false))[svc](grpcLocation, credentials, {
          "grpc.max_receive_message_length": 1024 * 1024 * 1000,
          "grpc.max_send_message_length": 1024 * 1024 * 1000
})`

Change 1000 to whatever you want, this would allow for 1000mb messages which may be excessive.
